src/server/routes/entity/process-unified-form.ts

```diff
@@ -242,116 +173,39 @@ export function handleCreateMultipleEntities(
-			if (savedMainEntity.get('type') === 'Edition') {
-				await indexAutoCreatedEditionGroup(orm, savedMainEntity, transacting);
-			}
-			bbidMap[entityKey] = savedMainEntity.get('bbid');
-			savedMainEntities[entityKey] = savedMainEntity.toJSON();
-			// getNextRelationshipSet expects relationshipSet.id to exist
-			if (!isNew) {
-				_.set(savedMainEntities[entityKey], ['relationshipSet', 'id'], savedMainEntities[entityKey].relationshipSetId);
-			}
-		}
-		try {
-			// bookshelf's transaction have issue with Promise.All, refer https://github.com/bookshelf/bookshelf/issues/1498 for more details
-			await Object.keys(body).reduce((promise, entityKey) => promise.then(() => processEntity(entityKey)), Promise.resolve());
 
-			// adding relationship on newly created entites
-			await Object.keys(allRelationships).reduce((promise, entityId) => promise.then(() => processRelationship(
-				allRelationships[entityId], savedMainEntities[entityId], bbidMap, editorJSON.id, orm, transacting
-			)), Promise.resolve());
-
-			return savedMainEntities;
-		}
-		catch (err) {
-			log.error(err);
-			throw err;
-		}
-	});
-	const achievementPromise = entityEditPromise.then(
-		(entitiesJSON:Record<string, any>) => {
-			const entitiesAchievementsPromise = [];
-			for (const entityJSON of Object.values(entitiesJSON)) {
-				entitiesAchievementsPromise.push(processAchievement(orm, editorJSON.id, entityJSON));
+	const {body}: {body: Record<string, any>} = req;
+	const processedAchievements = [];
+	async function processEntity(entityKey:string) {
+		const entityForm = body[entityKey];
+		const entityType = _.upperFirst(entityForm.type);
+		const deriveProps = _.pick(entityForm, additionalEntityProps[_.camelCase(entityType)]);
+		const isNew = _.get(entityForm, '__isNew__', true);
+		currentEntity = null;
+		if (!isNew) {
+			currentEntity = await orm.func.entity.getEntity(orm, entityType, entityForm.id, getEntityRelations(entityType as EntityTypeString));
+			if (!currentEntity) {
+				throw new FormSubmissionError('Entity with given id not found');
 			}
-			return Promise.all(entitiesAchievementsPromise).catch(err => { throw err; });
 		}
-	);
+		const {bookshelf} = orm;
+		const entityEditPromise = bookshelf.transaction((transacting) =>
+			processSingleEntity(entityForm, isNew ? null : currentEntity, null, entityType, orm, editorJSON, deriveProps, false, transacting));
+
+		const achievementPromise = entityEditPromise.then(
+			(entityJSON) => processAchievement(orm, editorJSON.id, entityJSON)
+		);
+		processedAchievements.push(await achievementPromise);
+		return achievementPromise;
+	}
+	// create all entities
+	async function getAchievementPromise() {
+		await Object.keys(body).reduce((promise, entityKey) => promise.then(() => processEntity(entityKey)), Promise.resolve());
+		return processedAchievements;
+	}
 	return handler.sendPromiseResult(
 		res,
-		achievementPromise,
+		getAchievementPromise(),
```

Comment:
This should probably not be called here, instead passing the promise and letting `sendPromiseResult` resolve the promise safely.
```suggestion
		getAchievementPromise,
```

Reply:
But function does return promise that resolves to entity object, right?

Reply:
Maybe I'm misreading the code, but in short by calling the function here you're executing the promise (without it being in a try-catch block; any error will be uncaught and crash the server).
The idea with the `handler.sendPromiseResult` helper is that it will take care of `await`-ing the promise in a try-catch block and figure out what to do on success or fail.
An async function *is* essentially a promise, but does not necessarily *return* a promise. In this case it returns the `processedAchievements` array.

Reply:
We discussed on IRC, I was totally wrong and I apparently can't read code :p
sendPromiseResult awaits the promise but does not execute it